README
= googlestaticmap gem

This gem is on Rubygems, simply type "gem install googlestaticmap" to install it.

Available on github at http://github.com/brentsowers1/googlestaticmap

Class for generating URLs for and downloading static maps from the Google Maps
Static API service.  GoogleStaticMap is the main class to use, instantiate it,
set the attributes to what you want, and call url on the instance to get the
URL to download the map from.  You can also call get_map to actually
download the map from Google, return the bytes, and optionally write the
image to a file.

== Examples:

Get a simple static map centered at Washington, DC, with the default size
(500 x 350), zoomed to level 11.  image will be the binary data of the map
  require 'googlestaticmap'
  map = GoogleStaticMap.new(:zoom => 11, :center => MapLocation.new(:address => "Washington, DC"))
  image = map.get_map

Get the URL of the image described in the previous example, so you can insert
this URL as the src of an img element on an HTML page
  require 'googlestaticmap'
  map = GoogleStaticMap.new(:zoom => 11, :center => MapLocation.new(:address => "Washington, DC"))
  image_url = map.url(:auto)

Get a map with blue markers at the White House and the Supreme Court, zoomed
the closest that the map can be with both markers visible, at the default
size.  image will be the binary data of the map
  require 'googlestaticmap'
  map = GoogleStaticMap.new
  map.markers << MapMarker.new(:color => "blue", :location => MapLocation.new(:address => "1600 Pennsylvania Ave., Washington, DC"))
  map.markers << MapMarker.new(:color => "blue", :location => MapLocation.new(:address => "1 1st Street Northeast, Washington, DC"))
  image = map.get_map

Get a GIF satellite map, with a size of 640 x 480, with a
semi transparent green box drawn around a set of 4 coordinates, with the box
outline solid, centered at the middle of the box, written out to the file
map.gif:
  require 'googlestaticmap'
  map = GoogleStaticMap.new(:maptype => "satellite", :format => "gif", :width => 640, :height => 480)
  poly = MapPolygon.new(:color => "0x00FF00FF", :fillcolor => "0x00FF0060")
  poly.points << MapLocation.new(:latitude => 38.8, :longitude => -77.5)
  poly.points << MapLocation.new(:latitude => 38.8, :longitude => -76.9)
  poly.points << MapLocation.new(:latitude => 39.2, :longitude => -76.9)
  poly.points << MapLocation.new(:latitude => 39.2, :longitude => -77.5)
  poly.points << MapLocation.new(:latitude => 38.8, :longitude => -77.5)
  map.paths << poly
  map.get_map("map.gif")

If you're working behind a proxy, create the map object this way:
  map = GoogleStaticMap.new(:proxy_address=>'my.proxy.host', :proxy_port=>8080, :width => 640, :height => 480)


Author:: Brent Sowers (mailto:brent@coordinatecommons.com)
License:: You're free to do whatever you want with this

To post comments about this gem, go to my blog posting at
http://www.brentsowers.com/2010/08/gem-for-getting-google-static-maps.html

History.txt
== 1.2.6 / 2018-12-08
* Added the visible property to set map locations that will be visible (thanks luca!)
* Changed default protocol to https
* Updated example script to ask for an API key
* Forced specification of an API key or client ID/private key, since Google now requires this
* Updated documentation to specify that an API key or client ID/private key is required

== 1.2.5 / 2018-05-07
* Added styles, plain_string options to GoogleStaticMap (thanks anplug!)
* Added anchor option to MapMarker (thanks anplug!)

== 1.2.4 / 2017-03-05
* Added language option (thanks kntmrkm!)

== 1.2.3 / 2017-01-12
* Signature is now sent when using an API key (thanks mauu-alpha!)

== 1.2.2 / 2014-03-08
* Added channel parameter for tracking (thanks dfenniak!)

== 1.2.1 / 2014-03-03
* Fixed bug in the signature generation for business customers from 1.2.0 (thanks nathany!)

== 1.2.0 / 2014-03-03
* Added support for setting an API key for monitoring, and a client ID/private key for business customers (thanks forrest!)
* Changed to use maps.googleapis.com instead of maps.google.com (thanks nathany!)
* Explicitly added MIT licensing

== 1.1.4 / 2013-07-04
* Properly URL encoding the URLs, caused an issue when using URI.parse on generated URLs (thanks nathany!)

=== 1.1.3 / 2012-07-08
* Fixed a Ruby 1.9.3 compatibility issue (thanks dusanb!)
* Fixed a Ruby 1.8.7 compatibility issue introduced in 1.1.1

=== 1.1.2 / 2012-04-27
* Fixed bug from 1.1.1 when specifying https when calling get_map (thanks aduncan for catching it!)

=== 1.1.1 / 2012-04-22
* Added support for specifying the protocol to use (thanks andruby!)

=== 1.1.0 / 2011-10-18
* Added an option to make requests through a proxy server (thanks bardsley!)

=== 1.0.1 / 2011-09-23
* Added a scale parameter to the map (thanks backspace!)

=== 1.0.0 / 2011-07-21
* An explicitly set map center is now properly encoded - was previously being doubly encoded (thanks abecciu)
* Gem will now load on some bad Rubygems installations (path issue)
* Version 1.0, yeah!

=== 0.1.5 / 2011-02-08
* Fixed bug which caused require 'googlestaticmap' to not work on some versions of Ruby

=== 0.1.4 / 2010-08-21
* Added unit tests

=== 0.1.2 / 2010-08-20
* Now works in Ruby 1.8 as well as 1.9
* Fixed a bug when getting a map without specifying a file

=== 0.1.0 / 2010-08-19
* Initial release


test/tc_helper.rb
require 'test/unit'
require 'cgi'
require File.dirname(__FILE__) +  '/../lib/googlestaticmap_helper'

class GoogleStaticMapHelperTest < Test::Unit::TestCase #:nodoc: all

  IVS = {"@var1" => 1, "@var2" => 2, "@var3" => 3,
         "@url" => "http://maps.google.com", "@nilvar" => nil}

  class MockRuby18Object
    def self.instance_variables
      IVS.keys
    end
    def self.instance_variable_get(variable)
      IVS[variable]
    end
  end

  # Ruby 1.9 refers to instance variables as symbols, like :@var1.  The
  # helper class needs to handle both cases
  class MockRuby19Object
    def self.instance_variables
      IVS.keys.collect {|k| k.to_sym}
    end
    def self.instance_variable_get(variable)
      IVS[variable.to_s]
    end
  end


  def test_safe_instance_variables_no_params
    [MockRuby18Object, MockRuby19Object].each do |o|
      sivs = GoogleStaticMapHelpers.safe_instance_variables(o)
      assert_equal ivs_no_at, sivs
      assert !sivs.has_key?("@nilvar")
    end
  end

  def test_safe_instance_variables_exclude
    [MockRuby18Object, MockRuby19Object].each do |o|
      sivs = GoogleStaticMapHelpers.safe_instance_variables(o, ["var2"])
      assert_equal IVS.length-2, sivs.length
      assert !sivs.has_key?("@var2")
    end
  end

  def test_safe_instance_variables_cgi
    [MockRuby18Object, MockRuby19Object].each do |o|
      sivs = GoogleStaticMapHelpers.safe_instance_variables(o, [], :cgi_escape_values => true)
      assert_equal IVS.length-1, sivs.length
      assert_equal CGI.escape(IVS["@url"]), sivs["url"]
    end
  end

  def test_signature
    # This comes from the google example at
    # https://developers.google.com/maps/documentation/business/webservices/auth
    private_key = "vNIXE0xscrmjlyV-12Nj_BvUPaw="
    url = "/maps/api/geocode/json?address=New+York&sensor=false&client=clientID"
    sig = GoogleStaticMapHelpers.sign(url, private_key)
    assert_equal sig, "KrU1TzVQM7Ur0i8i7K3huiw3MsA="
  end

  private
  def ivs_no_at
    ivs = {}
    IVS.each do |k,v|
      ivs[k[1..-1]] = v unless v.nil?
    end
    ivs
  end
end


test/tc_map_marker.rb
require 'test/unit'
require 'cgi'
require File.dirname(__FILE__) +  '/../lib/googlestaticmap'

class MapMarkerTest < Test::Unit::TestCase #:nodoc: all
  def test_exception_no_location
    m = MapMarker.new
    assert_raise Exception do
      m.to_s
    end
  end

  def test_exception_invalid_location
    m = MapMarker.new
    m.location = 99
    assert_raise Exception do
      m.to_s
    end
  end

  def test_set_parameters
    m = nil
    assert_nothing_raised { m = default_marker }
    assert_equal "green", m.color
    assert_equal "Washington, DC", m.location.address
    assert_equal "tiny", m.size
    assert_equal "B", m.label
    assert_equal "http://www.google.com", m.icon
    assert_equal false, m.shadow
    assert_equal 'center', m.anchor
  end

  def test_get_string
    m = default_marker
    s = nil
    assert_nothing_raised {s = m.to_s}
    assert_equal 7, s.split(MAP_SEPARATOR).length
    assert s.include?(CGI.escape("Washington, DC"))
    assert s.include?("color:green")
    assert s.include?("icon:http://www.google.com")
    assert s.include?("anchor:center")
  end

  def test_string_is_a_valid_ruby_uri
    m = default_marker
    URI.parse(m.to_s)
  end

  private
  def default_marker
    MapMarker.new(:color => "green",
                  :location => MapLocation.new(:address => "Washington, DC"),
                  :size => "tiny",
                  :label => "B",
                  :icon => "http://www.google.com",
                  :anchor => 'center',
                  :shadow => false)
  end
end


test/tc_map_location.rb
require 'test/unit'
require 'uri'
require File.dirname(__FILE__) +  '/../lib/googlestaticmap'

class MapLocationTest < Test::Unit::TestCase #:nodoc: all
  def test_exception_no_location
    m = MapLocation.new
    assert_raise Exception do
      m.to_s
    end
  end

  def test_attributes_filled_in
    m = MapLocation.new(:latitude => 39, :longitude => -77)
    assert_equal "39", m.latitude
    assert_equal "-77", m.longitude
    assert_nil m.address
    m = MapLocation.new(:address => "Washington, DC")
    assert_equal "Washington, DC", m.address
    assert_nil m.latitude
    assert_nil m.longitude
  end

  def test_with_address
    m = MapLocation.new(:address => "Washington, DC")
    s = ""
    assert_nothing_raised { s = m.to_s }
    assert_equal s, URI.escape("Washington, DC")
  end

  def test_with_latitude_longitude
    m = MapLocation.new(:latitude => 39, :longitude => -77)
    s = ""
    assert_nothing_raised { s = m.to_s }
    assert_equal s, "39,-77"
  end
end


example.rb
require File.dirname(__FILE__) + '/lib/googlestaticmap'

print "This script requires an API key to test. If you do not have one, visit\n"
print "https://developers.google.com/maps/documentation/javascript/get-api-key\n"
print "and click Get Started\n\n"
print "Please enter your API key: "
api_key = gets.chomp

print "-----------------------------------------------------------------------------\n"
print "--------------  Test 1, grabbing a map of Washington, DC --------------------\n"
print "-----------------------------------------------------------------------------\n"
print "\n"
map = GoogleStaticMap.new(
    :api_key => api_key, 
    :zoom => 11, 
    :center => MapLocation.new(:address => "Washington, DC"))
image = map.get_map
print "Got a map of size #{image.length}, should be roughly 60K\n"
image_url = map.url
print "The URL for this map is '#{image_url}'\n"


print "\n\n\n"
print "-----------------------------------------------------------------------------\n"
print "- Test 2, Get a map with blue markers at the White House and the Supreme    -\n" 
print "- Court, zoomed the closest that the map can be with both markers visible,  -\n"
print "- at the default size.                                                      -\n"
print "-----------------------------------------------------------------------------\n"
print "\n"
map = GoogleStaticMap.new(:api_key => api_key)
map.markers << MapMarker.new(:color => "blue", :location => MapLocation.new(:address => "1600 Pennsylvania Ave., Washington, DC"))
map.markers << MapMarker.new(:color => "blue", :location => MapLocation.new(:address => "1 1st Street Northeast, Washington, DC"))
image = map.get_map

print "Got a map of size #{image.length}, should be roughly 50K\n"
image_url = map.url
print "The URL for this map is '#{image_url}'\n"

print "\n\n\n"
print "-----------------------------------------------------------------------------\n"
print "- Test 3, Get a GIF satellite map, with a size of 640 x 480, with a semi    -\n"
print "- transparent green box drawn around a set of 4 coordinates, with the box   -\n"
print "- outline solid, centered at the middle of the box, written out to the file -\n"
print "- map.gif:                                                                  -\n"
print "-----------------------------------------------------------------------------\n"
print "\n"
map = GoogleStaticMap.new(:maptype => "satellite", :format => "gif", :width => 640, :height => 480,
                          :api_key => api_key)
poly = MapPolygon.new(:color => "0x00FF00FF", :fillcolor => "0x00FF0060")
poly.points << MapLocation.new(:latitude => 38.8, :longitude => -77.5)
poly.points << MapLocation.new(:latitude => 38.8, :longitude => -76.9)
poly.points << MapLocation.new(:latitude => 39.2, :longitude => -76.9)
poly.points << MapLocation.new(:latitude => 39.2, :longitude => -77.5)
poly.points << MapLocation.new(:latitude => 38.8, :longitude => -77.5)
map.paths << poly
map.get_map("map.gif")

print "There should now be a map.gif file in this directory.\n"




lib/googlestaticmap_helper.rb
require 'base64'
require 'openssl'

# Helper methods used in this code
module GoogleStaticMapHelpers #:nodoc: all
  # Returns a hash of instance variables for the passed in Object, where the
  # key is the instance variable name ("@instance_variable"), and the value is
  # the current instance variable value.  This is safe for Ruby 1.8 and 1.9.
  # Pass an optional array of instance variable name strings
  # (["instance_variable_1", "instance_variable_2"]) to not include in the
  # result.  Example result:
  # {"variable_1" => 123, "variable_2" => "2 value"}
  # All inputs and outputs do NOT have the @ in the instance variable name
  # options -
  #   :cgi_escape_values - set this to true to return all values as
  #   CGI escaped strings
  def self.safe_instance_variables(object, ivs_to_exclude=[], options={})
    ivs = {}
    object.instance_variables.each do |i|
      # Don't include the @
      iv_name = i.to_s[1..-1]
      unless ivs_to_exclude.include?(iv_name)
        val = object.instance_variable_get(i)
        unless val.nil?
          val = CGI.escape(val.to_s) if options.has_key?(:cgi_escape_values)
          ivs[iv_name] = val unless val.nil?
        end
      end
    end
    ivs
  end

  # signing code is grabbed from https://github.com/alexreisner/geocoder
  def self.sign(path, key)
    raw_private_key = self.url_safe_base64_decode(key)
    digest = OpenSSL::Digest.new('sha1')
    raw_signature = OpenSSL::HMAC.digest(digest, raw_private_key, path)
    self.url_safe_base64_encode(raw_signature)
  end

  def self.url_safe_base64_decode(base64_string)
    Base64.decode64(base64_string.tr('-_', '+/'))
  end

  def self.url_safe_base64_encode(raw)
    Base64.encode64(raw).tr('+/', '-_').strip
  end

end


lib/googlestaticmap.rb
# Main file for the googlestaticmap gem.  See README.md for a full desciption with examples, licensing, contact
# info, etc.

require 'cgi'
require 'net/http'
require 'net/https' if RUBY_VERSION < "1.9"
require File.dirname(__FILE__) +  '/googlestaticmap_helper'

MAP_SEPARATOR = CGI.escape("|")

# Main class for creating a static map.  Create an instance, Set attributes
# that describe properties of the map.  Then call url to get a URL that you
# can use as the src of an img tag.  You can also call get_map to actually
# download the map from Google, and optionally write it to a file.
class GoogleStaticMap
  # Width of resulting image in pixels, defaults to 500, maximum 640
  attr_accessor :width

  # Height of resulting image in pixels, defaults to 350, maximum 640
  attr_accessor :height

  # An optional array of MapMarker instances
  attr_accessor :markers

  # An optional array of MapLocation instances that should remain visible on the map, though no markers or other indicators will be displayed
  attr_accessor :visible

  # An optional array of MapPath instances and/or MapPolygon instances to draw
  attr_accessor :paths

  # MapLocation for the center of the map. If this is not specified, the map
  # will zoom to the markers
  attr_accessor :center

  # 0 (the whole world) to 21 (individual buildings)
  attr_accessor :zoom

  # Applications that determine the user's location via a sensor must set this
  # to true, defaults to false
  attr_accessor :sensor

  # 1, 2, or 4 for Maps API Premier customers. Defaults to 1. Makes everything
  # in the image appear larger, useful for displaying on high res mobile
  # screens.  When setting this, the image's actual width and height in pixels
  # will be scale * width and scale * height
  attr_accessor :scale

  # format of the image:
  # * png8 - 8 bit PNG (default)
  # * png32 - 32 bit PNG
  # * gif
  # * jpg
  # * jpg-baseline - non-progressive JPEG
  attr_accessor :format

  # Type of map to create:
  # * roadmap (default)
  # * satellite
  # * terrain
  # * hybrid - satellite imagery with roads
  attr_accessor :maptype

  # If you need to use a proxy server to reach Google, set the name/address
  # of the proxy server here
  attr_accessor :proxy_address

  # If proxy_address is set, set this to the port of the proxy server
  attr_accessor :proxy_port

  # API Key - see https://developers.google.com/maps/documentation/static-maps/get-api-key#key for details
  # Note that if this is set, you cannot provide a client ID
  attr_accessor :api_key

  # ClientId for business customers -
  # see https://developers.google.com/maps/documentation/static-maps/get-api-key#key for details
  # Note that if this is set, you cannot provide an API key.
  attr_accessor :client_id

  # The private key, also known as the URL signing secret, is used to to generate the signature parameter
  # in the URL. This is required if you are using a client ID, or a premium API key, and is optional
  # if you are using a standard API key.  See
  # https://developers.google.com/maps/documentation/static-maps/get-api-key for more details
  attr_accessor :private_key

  # Channel - identifier channel for tracking API source in enterprise tools
  #           see https://developers.google.com/maps/documentation/business/clientside/quota for details
  attr_accessor :channel

  # Language - :en, :ja
  #           see https://developers.google.com/maps/documentation/static-maps/intro for details
  attr_accessor :language

  # Styles - see https://developers.google.com/maps/documentation/maps-static/styling
  # styles is should be represented as array of objects
  # [{feature: 'featureArgument, element: 'elementArgument', rule1: 'rule1Arguement', rule2: 'rule2Arguement', ...},
  # ...]
  attr_accessor :styles

  # In case when some parameter should be inserted manually
  # For example, using https://mapstyle.withgoogle.com/ tool and inserting generated style as is
  attr_accessor :plain_string

  # Takes an optional hash of attributes
  def initialize(attrs={})
    defaults = {:width => 500, :height => 350, :markers => [], :visible => [],
                :sensor => false, :maptype => "roadmap", :paths => [],
                :proxy_port => nil, :proxy_address => nil, :api_key => nil,
                :client_id => nil, :private_key => nil, :language => nil}

    attributes = defaults.merge(attrs)
    attributes.each {|k,v| self.send("#{k}=".to_sym,v)}
  end

  # Returns the full URL to retrieve this static map.  You can use this as the
  # src for an img to display an image directly on a web page
  # Example - "https://maps.googleapis.com/maps/api/staticmap?params..."
  # +protocol+ can be 'http', 'https' or :auto. Specifying :auto will not return
  #   a protocol in the URL ("//maps.googleapis.com/..."), allowing the browser to
  #   select the appropriate protocol (if the page is loaded with https, it will
  #   use https). Defaults to https
  def url(protocol='https')
    unless @center || @markers.length > 0 || @paths.length > 0 || @visible.length > 0
      raise Exception.new("Need to specify either a center, markers, visible points, or a path")
    end
    if !@api_key.nil? && !@client_id.nil?
      raise Exception.new("You cannot specify both an API key and a client ID, only specify one")
    end
    if @api_key.nil? && @client_id.nil?
      raise Exception.new("You must specify either an API key, or a client ID/private key. Google requires this and the calls to Google will fail without one of these.")
    end
    if !@client_id.nil? && @private_key.nil?
      raise Exception.new("private_key must be specified if using a client ID")
    end
    protocol = 'https' unless protocol == 'http' || protocol == 'https' ||
                              protocol == :auto
    protocol = protocol == :auto ? '' : protocol + ":"
    base = "#{protocol}//maps.googleapis.com"
    path = "/maps/api/staticmap?"
    attrs = GoogleStaticMapHelpers.safe_instance_variables(self,
              ["markers", "visible", "paths", "width", "height", "center",
               "proxy_address", "proxy_port", "api_key", "client_id",
               "private_key", "styles", "plain_string"],
              :cgi_escape_values => true).to_a
    attrs << ["size", "#{@width}x#{@height}"] if @width && @height
    @markers.each {|m| attrs << ["markers",m.to_s] }
    @paths.each {|p| attrs << ["path",p.to_s] }
    attrs << ["visible", @visible.join(MAP_SEPARATOR)] if !@visible.nil? && @visible.any?
    attrs << ["center", @center.to_s] if !@center.nil?
    attrs << ["key", @api_key] if !@api_key.nil?
    attrs << ["client", @client_id] if !@client_id.nil?
    get_styles.each { |style| attrs << style } if !@styles.nil?
    path << attrs.sort_by {|k,v| k}.collect {|attr| "#{attr[0]}=#{attr[1]}"}.join("&")
    path << "&#{@plain_string}" if !@plain_string.nil?
    if (!@api_key.nil? || !@client_id.nil?) && !@private_key.nil?
      signature = GoogleStaticMapHelpers.sign(path, @private_key)
      path << "&signature=" << signature
    end
    base + path
  end

  # Returns the URL to retrieve the map, relative to https://maps.googleapis.com
  # Example - "/maps/api/staticmap?params..."
  def relative_url(protocol='https')
    url(protocol).gsub(/[^\/]*\/\/maps\.googleapis\.com/, "")
  end


  # Connects to Google, retrieves the map, and returns the bytes for the image.
  # Optionally, pass it an output name and the contents will get written to
  # this file name
  # +output_file+ - optionally give the name of a file to write the output to.
  #                 Pass nil to not write the output to a file
  # +protocol+ - specify http or https here for the protocol to retrieve the
  #              map with. Defaults to http
  # return value - the binary data for the map
  def get_map(output_file=nil, protocol='https')
    protocol = 'https' unless protocol == 'http' || protocol == 'https'
    port = protocol == 'https' ? 443 : 80
    http = Net::HTTP.Proxy(@proxy_address,@proxy_port).new("maps.googleapis.com", port)
    http.use_ssl = protocol == 'https'

    resp = http.get2(relative_url(protocol))
    if resp && resp.is_a?(Net::HTTPSuccess)
      if output_file
        File.open(output_file, "wb") {|f| f << resp.body }
      end
      resp.body
    else
      if resp
        raise Exception.new("Error encountered while retrieving google map, code #{resp.code}, text #{resp.body}")
      else
        raise Exception.new("Error while retrieve google map, no response")
      end
    end
  end

  def get_styles
    @styles.map do |style|
      values = style.each_pair.map do |(key, value)|
        "#{key.to_s}:#{value}"
      end
      ["style", values.join(CGI.escape('|'))]
    end
  end
end

# Container class for a location on the map.  Set either a latitude and
# longitude, or an address
class MapLocation
  # Decimal degrees, positive is north
  attr_accessor :latitude

  # Decimal degrees, positive is east
  attr_accessor :longitude

  # String address - can be a full address, city name, zip code, etc.  This is
  # ignored if latitude and longitude are set.
  attr_accessor :address

  def initialize(attrs={})
    attrs.each {|k,v| self.send("#{k}=".to_sym,v.to_s)}
  end

  def to_s
    if latitude && longitude
      "#{CGI.escape(latitude.to_s)},#{CGI.escape(longitude.to_s)}"
    elsif address
      CGI.escape(address)
    else
      raise Exception.new("Need to set either latitude and longitude, or address")
    end
  end
end

# A single marker to place on the map.  Initialize and pass
class MapMarker
  # A 24-bit color (example: 0xFFFFCC) or a predefined color from the set
  # {black, brown, green, purple, yellow, blue, gray, orange, red, white}
  attr_accessor :color

  # MapLocation for the position of this marker
  attr_accessor :location

  # (optional) The size of marker from the set {tiny, small, mid}. Defaults to
  # mid
  attr_accessor :size

  # (optional) A single uppercase alphanumeric character label from the set
  # {A-Z, 0-9}.  Not allowed for tiny and small sizes
  attr_accessor :label

  # a URL to use as the marker's custom icon. Images may be in PNG, JPEG or GIF
  # formats, though PNG is recommended
  attr_accessor :icon

  # true (default) indicates that the Static Maps service should construct an
  # appropriate shadow for the image. This shadow is based on the image's
  # visible region and its opacity/transparency.
  attr_accessor :shadow

  # Manage position of marker (anchor:center/bottom/top..)
  # See - https://developers.google.com/maps/documentation/maps-static/intro#Markers
  attr_accessor :anchor

  # Takes an optional hash of attributes
  def initialize(attrs={})
    attrs.each {|k,v| self.send("#{k}=".to_sym,v)}
  end

  def to_s
    raise Exception.new("Need a location for the marker") unless @location && @location.is_a?(MapLocation)
    attrs = GoogleStaticMapHelpers.safe_instance_variables(self, ["location"])
    s = attrs.to_a.sort_by {|x| x[0]}.collect do |k|
      # If the icon URL is URL encoded, it won't work
      val = (k[0] == "icon" ? k[1] : CGI.escape(k[1].to_s))
      "#{k[0]}:#{val}"
    end.join(MAP_SEPARATOR)
    s << MAP_SEPARATOR << @location.to_s
  end
end

# A path line to draw on the map.  Initialize and set attributes.
class MapPath
  # Thickness of the path line in pixels, defaults to 5
  attr_accessor :weight

  # Color of the path, either as a 24-bit (example: color=0xFFFFCC), a 32-bit
  # hexadecimal value (example: color=0xFFFFCCFF), or from the set
  # {black, brown, green, purple, yellow, blue, gray, orange, red, white}
  # When a 32-bit hex value is specified, the last two characters specify the
  # 8-bit alpha transparency value. This value varies between 00
  # (completely transparent) and FF (completely opaque).
  attr_accessor :color

  # MapPositions for each point on the line
  attr_accessor :points

  # Pass an optional hash of arguments
  def initialize(attrs={})
    @points = []
    attrs.each {|k,v| self.send("#{k}=".to_sym,v)}
  end

  def to_s
    raise Exception.new("Need more than one point for the path") unless @points && @points.length > 1
    attrs = GoogleStaticMapHelpers.safe_instance_variables(self, ["points"])
    s = attrs.to_a.sort_by {|x| x[0]}.collect {|k| "#{k[0]}:#{CGI.escape(k[1].to_s)}"}.join(MAP_SEPARATOR)
    s << MAP_SEPARATOR << @points.join(MAP_SEPARATOR)
  end
end

# A polygon to draw and fill on the map.  Has the same properties as a path
# with the addition of a fill color.  The points do not need to form a closed
# loop, the last point will automatically be joined the first point
class MapPolygon < MapPath
  # Color to fill in, either as a 24-bit (example: color=0xFFFFCC), a 32-bit
  # hexadecimal value (example: color=0xFFFFCCFF), or from the set
  # {black, brown, green, purple, yellow, blue, gray, orange, red, white}
  # When a 32-bit hex value is specified, the last two characters specify the
  # 8-bit alpha transparency value. This value varies between 00
  # (completely transparent) and FF (completely opaque).
  attr_accessor :fillcolor

  def initialize(attrs={})
    @points = []
    attrs.each {|k,v| self.send("#{k}=".to_sym,v)}
  end
end
